scripts/main.js

```diff
@@ -0,0 +1,348 @@
+        let newY = y;
+        let moveDistance = 0; // Variable to track the distance the cell will move
+
+        // Move the cell up as far as possible
+        while (newY > 0 && board[newY - 1][x] === 0) {
+          newY--;
+          moveDistance++; // Increment move distance
+        }
+
+        // Check if we can merge the cell with the one above
+        if (
+          newY > 0 &&
+          board[newY - 1][x] === board[y][x] &&
+          !merged[newY - 1]
+        ) {
+          // Merge the cells by doubling the value
+          board[newY - 1][x] *= 2;
+          scoreVal += board[newY - 1][x]; // Increase the score
+          board[y][x] = 0; // Clear the original cell
+          merged[newY - 1] = true; // Mark this cell as merged
+          hasChanged = true;
+
+          // Animate the move with merging (both cells move to the same position)
+          const cellElement = rows[y].cells[x].firstChild;
+          if (cellElement) {
+            cellElement.classList.add(`move-up--${moveDistance + 1}`); // Add a class for animation
+          }
+        } else if (newY !== y) {
+          // If the cell moved but did not merge
+          board[newY][x] = board[y][x]; // Move the cell to its new position
+          board[y][x] = 0; // Clear the original position
+          hasChanged = true;
+
+          // Animate the move without merging
+          const cellElement = rows[y].cells[x].firstChild;
+          if (cellElement) {
+            cellElement.classList.add(`move-up--${moveDistance}`); // Add a class for the move animation
+          }
+        }
+      }
+    }
+  }
+
+  return hasChanged; // Return whether the board changed
+}
+
+function moveDown() {
+  const originalBoard = board.map((row) => row.slice());
+  let hasChanged = false;
+
+  for (let x = 0; x < 4; x++) {
+    let merged = [false, false, false, false];
+
+    for (let y = 2; y >= 0; y--) {
+      if (board[y][x] !== 0) {
+        let newY = y;
+        let moveDistance = 0;
+
+        while (newY < 3 && board[newY + 1][x] === 0) {
+          newY++;
+          moveDistance++;
+        }
+
+        if (
+          newY < 3 &&
+          board[newY + 1][x] === board[y][x] &&
+          !merged[newY + 1]
+        ) {
+          board[newY + 1][x] *= 2;
+          scoreVal += board[newY + 1][x];
+          board[y][x] = 0;
+          merged[newY + 1] = true;
+          hasChanged = true;
+
+          const cellElement = rows[y].cells[x].firstChild;
+          if (cellElement) {
+            cellElement.classList.add(`move-down--${moveDistance + 1}`);
+          }
+        } else if (newY !== y) {
+          board[newY][x] = board[y][x];
+          board[y][x] = 0;
+          hasChanged = true;
+
+          const cellElement = rows[y].cells[x].firstChild;
+          if (cellElement) {
+            cellElement.classList.add(`move-down--${moveDistance}`);
+          }
+        }
+      }
+    }
+  }
+
+  return hasChanged;
```

Comment:
The 'moveDown' function is very similar to 'moveUp', 'moveLeft', and 'moveRight'. These functions share a lot of logic, which suggests that you could create a single, more generic move function that takes direction as a parameter to reduce code repetition.